Libraries/LibC/spawn.cpp

```diff
@@ -117,33 +131,42 @@ int posix_spawnp(pid_t* out_pid, const char* path, const posix_spawn_file_action
     posix_spawn_child(path, file_actions, attr, argv, envp, execvpe);
 }
 
-
-#if 0
-// FIXME
-int posix_spawn_file_actions_addclose(posix_spawn_file_actions_t*,
-    int)
+int posix_spawn_file_actions_addclose(posix_spawn_file_actions_t* actions, int fd)
 {
+    actions->state->actions.append([fd]() { return close(fd); });
+    return 0;
 }
 
-int posix_spawn_file_actions_adddup2(posix_spawn_file_actions_t*,
-    int, int)
+int posix_spawn_file_actions_adddup2(posix_spawn_file_actions_t* actions, int old_fd, int new_fd)
 {
+    actions->state->actions.append([old_fd, new_fd]() { return dup2(old_fd, new_fd); });
+    return 0;
 }
 
-int posix_spawn_file_actions_addopen(posix_spawn_file_actions_t*,
-    int, const char*, int, mode_t)
+int posix_spawn_file_actions_addopen(posix_spawn_file_actions_t* actions, int want_fd, const char* path, int flags, mode_t mode)
 {
+    actions->state->actions.append([want_fd, path, flags, mode]() {
+        int opened_fd = open(path, flags, mode);
+        if (opened_fd < 0 || opened_fd == want_fd)
+            return opened_fd;
+        if (int rc = dup2(opened_fd, want_fd); rc < 0)
+            return rc;
+        return close(opened_fd);
+    });
+    return 0;
 }
 
-int posix_spawn_file_actions_destroy(posix_spawn_file_actions_t*)
+int posix_spawn_file_actions_destroy(posix_spawn_file_actions_t* actions)
 {
+    delete actions->state;
+    return 0;
 }
 
-int posix_spawn_file_actions_init(posix_spawn_file_actions_t*)
+int posix_spawn_file_actions_init(posix_spawn_file_actions_t* actions)
 {
+    actions->state = new posix_spawn_file_actions_state;
```

Comment:
Hm, I suppose this could throw. What's serenity's policy on OOM from alloc? Should I use `new (std::nothrow)` here? Does Vector have a nothrow append() version? Should wrap this in a try/catch(bad_alloc)? Or does oom just always kill the process?

Reply:
I'm pretty sure we don't have any exceptions (yet?), so oom would just crash the process.

Reply:
We don't use exceptions (by choice) :)
malloc() currently does not fail (it crashes instead.) If you want failable allocations, there's mmap().

Reply:
Ok, then I think this is should be fine as-is.